File: src/main/java/com/chaco/algorithms/array/spiralOrder.java
package com.chaco.algorithms.array;

import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

/**
 * <a href="https://leetcode.cn/problems/spiral-matrix/">...</a>
 * 给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。
 * 输入：matrix = [[1,2,3],[4,5,6],[7,8,9]]
 * 输出：[1,2,3,6,9,8,7,4,5]
 * <p>
 * 输入：matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
 * 输出：[1,2,3,4,8,12,11,10,9,5,6,7]
 */
public class spiralOrder {
    public static void main(String[] args) {
        int[][] arr = new int[][]{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        List<Integer> integers = spiralOrder(arr);
        System.out.println("spiralOrder.main" + integers);
    }

    public static List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new LinkedList<>();
        if (null == matrix || Objects.equals(0, matrix.length) || Objects.equals(0, matrix[0])) {
            return res;
        }
        int m = matrix.length, n = matrix[0].length;
        int left = 0, right = n - 1, top = 0, bottom = m - 1;
        while (top <= bottom && left <= right) {
            for (int i = left; i <= right; i++) {
                res.add(matrix[top][i]);
            }
            for (int i = top + 1; i <= bottom; i++) {
                res.add(matrix[i][right]);
            }
            if (left < right && top < bottom) {
                for (int i = right - 1; i >= left; i--) {
                    res.add(matrix[bottom][i]);
                }
                for (int i = bottom - 1; i > top; i--) {
                    res.add(matrix[i][left]);
                }
            }
            left++;
            right--;
            top++;
            bottom--;
        }
        return res;
    }
}

File: src/main/java/com/chaco/algorithms/array/spiralOrder.go
package array

import "fmt"

func spiralOrder(matrix [][]int) []int {
	if nil == matrix || 0 == len(matrix) || 0 == len(matrix[0]) {
		return []int{}
	}
	var (
		left   = 0
		top    = 0
		right  = len(matrix[0]) - 1
		bottom = len(matrix) - 1
		index  = 0
		res    = make([]int, len(matrix)*len(matrix[0]))
	)
	for left <= right && top <= bottom {
		for i := left; i <= right; i++ {
			res[index] = matrix[top][i]
			index++
		}
		for i := top + 1; i <= bottom; i++ {
			res[index] = matrix[i][right]
			index++
		}
		if left < right && top < bottom {
			for i := right - 1; i >= left; i-- {
				res[index] = matrix[bottom][i]
				index++
			}
			for i := bottom - 1; i > top; i-- {
				res[index] = matrix[i][left]
				index++
			}
		}
		left++
		right--
		top++
		bottom--
	}
	return res
}

func main() {
	var arr = [][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}
	order := spiralOrder(arr)
	fmt.Println(order)
}
